katsdpcontroller/product_config.py

```diff
@@ -374,6 +414,62 @@ def from_config(cls,
         )
 
 
+class NgcAntennaChannelisedVoltageStream(AntennaChannelisedVoltageStreamBase):
+    """Real antenna-channelised-voltage stream (next-gen correlator).
+
+    It currently only supports wideband.
+    """
+
+    stream_type: ClassVar[str] = 'ngc.antenna_channelised_voltage'
+    _valid_src_types: ClassVar[_ValidTypes] = {'sim.dig.raw_antenna_voltage'}
+
+    def __init__(self, name: str, src_streams: Sequence[Stream], *,
+                 n_chans: int) -> None:
+        if n_chans < 1 or (n_chans & (n_chans - 1)) != 0:
+            raise ValueError('n_chans is not a power of 2')
+        if len(src_streams) % 2 != 0:
+            raise ValueError('src_streams does not have an even number of elements')
+        first = src_streams[0]
+        assert isinstance(first, SimDigRawAntennaVoltageStream)
```

Comment:
I'm ... wary of `assert` statements anywhere except in unit tests.* AFAIK if you run Python with the `-O` option then the statement just gets ignored.

* Although you could argue that sprinkling them in your code is useful for testing, knowing that they will have zero impact on the actual production code.

Reply:
In general (and in many languages), `assert` statements in non-test code are for checking for conditions that indicate bugs (whereas exceptions generally indicate that the code throwing the exception has encountered an exceptional condition but is not necessarily buggy). They're turned off in optimised code because you wouldn't ship buggy code, right? :-) And then they help in debugging by exploding as soon as the bug is detected, rather than 10 steps later in some obscure way.

In this specific case, it's related to mypy. It recognises assert statements like this and after this line assumes that `first` is an instance of `SimDigRawAntennaVoltageStream` and can type-check it appropriately.

Reply:
Right. Python and its dynamic typing coming to create philosophical conundrums.

Your explanation makes sense. I'll allow it. I'll have to do some reading up.